core/storage/user/gae_models.py

```diff
@@ -44,36 +44,80 @@ class UserSettingsModel(base_models.BaseModel):
     Instances of this class are keyed by the user id.
     """
 
-    # User id used to identify user by GAE. Is not required for now because we
-    # need to perform migration to fill this for existing users.
-    gae_id = ndb.StringProperty(required=True, indexed=True)
-    # Email address of the user.
-    email = ndb.StringProperty(required=True, indexed=True)
+    # Attributes used for both full users and profile users.
+
     # User role. Required for authorization. User gets a default role of
     # exploration editor.
     # TODO(1995YogeshSharma): Remove the default value once the one-off
     # migration (to give role to all users) is run.
     role = ndb.StringProperty(
         required=True, indexed=True, default=feconf.ROLE_ID_EXPLORATION_EDITOR)
+    # When the user last agreed to the terms of the site. May be None.
+    last_agreed_to_terms = ndb.DateTimeProperty(default=None)
+    # When the user last logged in. This may be out-of-date by up to
+    # feconf.PROXIMAL_TIMEDELTA_SECS seconds.
+    last_logged_in = ndb.DateTimeProperty(default=None)
+    # Actual name of the user displayed on android. Unlike username, it can be
+    # edited and is unique only among the profiles of the account.
+    actual_name = ndb.StringProperty(default=None)
+    # A code associated with each user on android to provide a PIN based
+    # authentication to access a profile within the account. For a full user
+    # it provides access to administrator controls in the account (on android).
+    pin = ndb.StringProperty(default=None)
```

Comment:
Should this be a part of `UserAuthModel`, also should this be somehow hashed or encrypted?

Reply:
Thanks for your comment!!
I think there's a confusion regarding the purpose of UserAuthModel. The UserAuthModel is only meant for storing information(metadata like gae_id) related to actual login process which involves calling an actual authentication service and doing the auth.
Whereas PIN is meant for in-account login - we can call it as profile login. So it's not connected to email with which the account was created or that sort of information. Like even when you are using the app in an offline mode, you can use the PIN based authentication - I am not sure how that is actually implemented in the android app, but it's totally an in-app thing and does not call any external service.

Regarding hashing it, that's a good suggestion. But do we need to consider about that in the storage itself or can I deal with it in later PRs - i.e do we want a class method for it implemented here or can I just implement it in domain layer - hash the PIN before updating it in the storage? My feeling is I can do it in domain layer.

Reply:
I was actually assuming that the PIN _is_ the authentication method for profiles, or rather, a combination of the parent user ID + PIN. By that thinking, a profile authenticates by providing a valid parent user ID & a PIN.

Re: hashing, I think that's a good idea. We don't let you recover PINs, only reset them, so we definitely don't need to store the actual PIN server-side. I don't think it's quite as sensitive as a password, but general hashing seems like a good idea.

Reply:
@BenHenning A combination of user ID + PIN will only be used actually. But I don't think we need to treat PIN as an authentication information to be sent along with the request to the server, as I am not planning to verify PIN at the backend.

The PIN (or its hash) is stored locally as well as on servers in the user data. Local PIN based auth system will ensure that only the correct user is making changes to that particular profile. Once that verification is done locally, PIN is no more used. At the backend we won't really check the PIN of the profile while processing its request, just use gae_id to verify that the request is coming from the correct account. We just use the gae_id, user_id to complete the request. Android side API can make sure to send the correct user id only of the profile -- For this particular reason I don't think we need to store PIN in `UserAuthModel`.

Reply:
How do we validate that a profile has access to certain APIs without verifying that it's logged in? I understand that we'll also be checking the PIN locally, but don't we also need to remotely validate the user in order for it to be considered 'logged in' for the purpose of retrieving data?

Reply:
@BenHenning Yes your right, we must do that verification too. Although it doesn't matter where we keep PIN for that auth purpose, but keeping it in the `UserAuthModel` only makes more sense since its an auth detail. 
I got a little lost into the details how the actual auth flow would look like and all, so didn't focus on the actual thing you were talking about which is just how we store it.

Hashing on `PIN` has been added as a point in the issue #10052 dealing with domain layer changes.

Reply:
Done.

Reply:
Thanks!